File: CPlus/CourseSchedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>> &prerequisites) {
        map<int, vector<int>> graph;
        vector<int> color(numCourses);

        for(auto &p: prerequisites) {
            graph[p.second].push_back(p.first);
        }

        for(int n=0; n<numCourses; n++) {
            if(color[n] != 2 && isCyclic(n, graph, color,n))
                return false;
        }
        return true;
    }

    bool isCyclic(int n, map<int, vector<int>> &g, vector<int> &color, int start) {
        if(color[start] == 1)
            return true;
        if(color[start] == 2)
            return false;
        
        color[start] = 1;
        for(int n: g[start]) {
            if(color[n] != 2) {
                if(isCyclic(n, g, color, n))
                    return true;
            }
        }
        color[start] = 2;
        return false;
    }
};

File: Python/courseSchedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph={}
        indegree={}
        
        for i in range(numCourses):
            indegree[i]=0
            graph[i]=[]
        
        for i,j in prerequisites:
            graph[j].append(i)
            indegree[i]+=1
        
        count=0
        queue=collections.deque()
        
        for course in indegree:
            if indegree[course]==0:
                queue.append(course)
        
        while queue:
            node=queue.popleft()
            count+=1
            
            for nei in graph[node]:
                indegree[nei]-=1
                if indegree[nei]==0:
                    queue.append(nei)
        
        
        return count==numCourses